mechanicalsoup/stateful_browser.py

```diff
@@ -262,7 +266,34 @@ def follow_link(self, link=None, *args, **kwargs):
                     self.list_links()
                     self.launch_browser()
                 raise
-        return self.open_relative(link['href'])
+        url = link['href']
+        if return_url_only:
+            return self.absolute_url(url)
+        else:
+            return self.open_relative(url)
+
+    def download_link(self, filename, *args, **kwargs):
+        """Downloads the contents of a link to a file. The browser state
+        will not change when calling this function.
+
+        :param filename: Filesystem path where the page contents will be
+            downloaded.
+        :param \*args, \*\*kwargs: Arguments forwarded to :func:`follow_link`.
+            These specify a link to the page whose contents you want to
+            download.
+        """
+        url = self.follow_link(return_url_only=True, *args, **kwargs)
```

Comment:
As per my comment above, I would rather see this as something like the following to avoid adding an option to `follow_link` that feels unnatural to me:

```python
    url = self.find_link(*args, **kwargs)['href']
```
Doing this means `download_link` can't take a bs4 tag as the link, which I'm fine with; however, if we demand it, then I think we can just reuse (or factor out) the "link-or-tag" input logic from `follow_link`, e.g.:

```python
def download_link(self, filename, link=None, *args, **kwargs):
    if not hasattr(link, 'attrs') or 'href' not in link.attrs:
        link = self.find_link(link, *args, **kwargs)
    url = link['href']

Reply:
You'll also need to deal with failure cases (follow_link has a special case to launch a browser when the debug mode is active, which is really convenient).

My first version (which I didn't push) factored this logic in a `find_link_or_fail` method. But in the end, the code was split into too many trivial methods to my taste.